### DISCUSSION TITLE
macOS code signing for Bareos packages

### DISCUSSION THREAD

**Discussion by onedayfishsale** [bareos/bareos#1959] [Q&A]
This is a continuation of https://groups.google.com/g/bareos-users/c/m5bKZM803iA but I'll restate the important points here.

My org uses an MDM for macOS devices, and we've been updating the policy we have to permit bareos-fd to access the full filesystem when we upgrade the Bareos package. This policy uses the signature on the bareos-fd binary. I'm not sure how long we've had this policy in place, but when we went to update the policy after upgrading to Bareos 22.1.6, we found that the bareos-fd binary was not signed.

In the bareos-users discussion I posted examples for the 22.1.6 and 22.1.5 subscription binaries but I've done some additional checking to determine when this started. I checked all the 22.1.x and 23.0.x packages so far:

```
bareos-22.1.0/usr/local/bareos/sbin/bareos-fd: code object is not signed at all
bareos-22.1.1/usr/local/bareos/sbin/bareos-fd: code object is not signed at all
bareos-22.1.2/usr/local/bareos/sbin/bareos-fd: code object is not signed at all
bareos-22.1.3/usr/local/bareos/sbin/bareos-fd: code object is not signed at all
Executable=/tmp/bareos-22.1.4/usr/local/bareos/sbin/bareos-fd
# designated => cdhash H"cbc07f18830e25e5b060c844885cb847fc36ac85"
Executable=/tmp/bareos-22.1.5/usr/local/bareos/sbin/bareos-fd
# designated => cdhash H"65a23b668a41c7f2d56f03d47bf6e517877090e8"
bareos-22.1.6/usr/local/bareos/sbin/bareos-fd: code object is not signed at all
bareos-23.0.0/usr/local/bareos/sbin/bareos-fd: code object is not signed at all
bareos-23.0.1/usr/local/bareos/sbin/bareos-fd: code object is not signed at all
bareos-23.0.2/usr/local/bareos/sbin/bareos-fd: code object is not signed at all
bareos-23.0.3/usr/local/bareos/sbin/bareos-fd: code object is not signed at all
bareos-23.0.4/usr/local/bareos/sbin/bareos-fd: code object is not signed at all
```

I used a quick script to do these checks:

```
#!/bin/sh

for pkg in bareos-*.pkg
do
  dir=$(basename $pkg .pkg)
  mkdir $dir
  xar -C $dir -xf $pkg
  gunzip < $dir/${dir}-filedaemon.pkg/Payload | (cd $dir && cpio -iud 2> /dev/null) 
  codesign --display --requirements - $dir/usr/local/bareos/sbin/bareos-fd
done
```

It seems like we must have started with our MDM policy at the time we were running 22.1.4, which just happened to be the first version (that we ran) that was signed.

So my question is: what is the intention for signing macOS binaries? My preference would be that they're signed, but either way is there maybe some inconsistency in the build workflow?

**Comment by arogge** [comment]
Sorry for taking that long. We discussed this issue and we all agree that we never signed any of the MacOS programs.
Could you try to find the certificate that signed the programs you have (codesign should be able to dump that somehow).
Thanks!

**Reply by onedayfishsale** [reply]
Thanks for the reply. I wasn't familiar with macOS code signing beyond the command our macOS folks gave me, but after doing some more reading I now see these use ad-hoc code signing. So there's no certificate involved here at all. See the `Signature=` line:

```
% codesign -d -vvv bareos-22.1.5/usr/local/bareos/sbin/bareos-fd
Executable=/tmp/bareos-22.1.5/usr/local/bareos/sbin/bareos-fd
Identifier=bareos-fd-55554944850582ca859a31d3ad3d54cf96c4956a
Format=Mach-O thin (x86_64)
CodeDirectory v=20400 size=17803 flags=0x2(adhoc) hashes=550+2 location=system
Hash type=sha256 size=32
CandidateCDHash sha256=65a23b668a41c7f2d56f03d47bf6e517877090e8
CandidateCDHashFull sha256=65a23b668a41c7f2d56f03d47bf6e517877090e83c63dd50801bfc9ea914591d
Hash choices=sha256
CMSDigest=65a23b668a41c7f2d56f03d47bf6e517877090e83c63dd50801bfc9ea914591d
CMSDigestType=2
CDHash=65a23b668a41c7f2d56f03d47bf6e517877090e8
Signature=adhoc
Info.plist=not bound
TeamIdentifier=not set
Sealed Resources=none
Internal requirements count=0 size=12
% 
```

For our MDM policy purposes, I believe we're just using that signature to verify the binary hasn't been modified or replaced, but not validating it against a certificate chain.

I learned more about ad-hoc signatures here:

https://apple.stackexchange.com/questions/288291/what-are-the-restrictions-of-ad-hoc-code-signing

Maybe there was a different build process used for 22.1.4 and 22.1.5 that resulted in ad-hoc signatures being applied to the bareos-fd binaries? Either way, for our specific purpose I think we can apply ad-hoc (or real) signatures to the newer versions ourselves but I'll check with the macOS folks here to make sure.

**Reply by onedayfishsale** [reply]
Quick update on this. We just got 22.1.6 signed and installed. The signing can be done with an ad-hoc code signing certificate:

`sudo codesign --force --verbose -s "My Org" -i org.example.bareos-fd /usr/local/bareos/sbin/bareos-fd​​​`

I'm not sure on the details but have seen a couple of mentions of macOS Sequoia being more restrictive when it comes to unsigned packages (or binaries, not sure which). So publishing signed packages/binaries might something the Bareos team wants to reconsider.
